File: leetcode/290.java
class Solution {
    public boolean wordPattern(String pattern, String str) {
        HashMap<String, Character> map = new HashMap<>();
        HashSet<Character> set = new HashSet<>();
        Character cVal;
        StringBuilder sb = new StringBuilder();
        int j = 0;
        str += " ";

        for (int i = 0; i < str.length(); i++) {
            if (str.charAt(i) == ' ') {
                cVal = map.get(sb.toString());
                if (j > pattern.length() - 1) return false;
                if (cVal == null) {
                    cVal = pattern.charAt(j);
                    if (true == set.contains(cVal)) return false;
                    set.add(cVal);
                    map.put(sb.toString(), cVal);
                } else {
                    if (false == cVal.equals(pattern.charAt(j)))
                        return false;
                }
                sb.delete(0, sb.length());
                j++;
            } else {
                sb.append(str.charAt(i));
            }
        }

        if (j != pattern.length()) return false;

        return true;
    }
}

File: leetcode/290.cpp
class Solution {
public:
	bool wordPattern(string pattern, string str) {
		unordered_map<string, char> map;
		unordered_set<char> set;
		string word;
		char cVal;
		int j = 0;
		str += " ";
		for (int i = 0; i < str.size(); i++) {

			if (str[i] == ' ') {
				if (j + 1 > pattern.size()) return false;

				if (map.find(word) != map.end()) {
					cVal = map[word];
					if (cVal != pattern[j]) return false;
				}
				else {
					cVal = pattern[j];
					if (set.find(cVal) != set.end()) return false;
					set.insert(cVal);
					map[word] = cVal;
				}
				word.clear();
				j++;
			}
			else {
				word += str[i];
			}
		}
		if (j < pattern.size()) return false;

		return true;
	}
};